docs/internal/types.md

```diff
@@ -0,0 +1,359 @@
+# Verona Type System
+
+## WARNING
+
+This is a preliminary document under heavy discussion.
+The ideas here are mostly right, the details, not necessarily.
+There will be a tutorial document that will be correct in all counts, but this isn't it.
+
+## Overview
+
+Verona is a statically typed language, with a structural and algebraic type system.
+
+The 'statically typed' bit means object types are always concrete and there is never implicit dynamic coercion like Perl or PHP.
+However, variable types can have a set of constraints instead of a concrete type, if and only if the possible types match the constraints.
+For example, variables can have abstract (interface) types if the objects bound to it are in a set of concrete (class) types that implement that interface.
+Another example is a type union, where a variable can be one of many types, but the code must deal with all possible types.
+The verification of these properties is done at compile time.
+
+The 'structural' bit means that open-world types can be defined by their shape: If an interface defines only a method `foo()` then any class that implements a method `foo()` implements this interface.
+This is in contrast to nominal typing, such as in Java, where an object implements an interface only if it adopts that interface.
+Objective-C supports both, with protocols defining nominal types but dynamic dispatch and explicit method checks providing structural typing.
+
+The 'algebraic' bit means that we can define types in terms of intersections or unions on other type expressions.
+For example, we can say `Integer | String` to define a set of constraints on a type that mean that it must either be an integer or a string (whatever they are: they may be concrete types, or they may be interfaces) or something like `Copyable & Hashable` to say any type that meets the constraints of the copyable and hashable interfaces.
+
+All types should be known at compile time and the compiler can assert that the objects and operations have the correct type at all times.
+Some types are inferred by the compiler, but there is no guessing: there's either one clear inferred type or it's an error.
```

Comment:
I'm not sure what you mean by this. @sylvanc I think you were talking about this a few months ago, any insight?

Reply:
Sorry, was too terse.  "having principal types" is a property of a type system, specifically, that every expression has exactly one most general type, in that every other type it can be said to have is a refinement of that most general type.  I was meaning to ask if this is true of Verona's type system in general (as is true of, say, SML), so there truly is only ever one answer (or the program is ill-typed), or if this is describing a policy of the type _inference_ system (as with Haskell with GADTs).  (Harper, who was my introduction to all this and so his vocabulary got stuck in my head, uses "elaboration" instead of "type inference", as [without putting too many words in his mouth, I hope] he wants people to think about it primarily as the transformation of a program with no or some type annotations into a program in which every term is annotated, with inference being a particular mechanism towards that end.)

Reply:
I may be wrong, but I believe Verona's type inference is a mechanism of validation, not of construction. So types are what they are and the inference only exposes the obvious, not creates new patterns. IIUC, this means Verona does have principal types. Though, this is something that Sylvan knows better, so we could just talk about this on our next meeting.